### ISSUE TITLE
Create the ability to toggle posts as pinned or unpinned

### ISSUE THREAD

**Issue by palisadoes** [PalisadoesFoundation/talawa-api#1096]
**Is your feature request related to a problem? Please describe.**

1. The Talawa home screen shows a newsfeed of  posts from Talawa users.
1. In the Talawa Admin console there are plans to have the ability for administrators to not only add posts to the news feed, but to also pin them to the top of the newsfeed
1. In anticipation of that feature we need the API to support this requirement
1. Look at the video below to see how the pinned posts are expected to be rendered on the Talawa mobile app screen

**Describe the solution you'd like**
1. Create ways to:
    1. store the pinned characteristic to every post
    1. update this characteristic to have one of two states, pinned or unpinned
    1. determine whether a post is pinned or not for the mobile app.
    1. create a way for the mobile app to get the N th. currently pinned post 
        1. always in reverse chronological order
        1. indicating whether the post is the last one in the list
1. 100% test code coverage 
1.  Update these Talawa-Docs  as necessary (This will require another issue created in that repo by you)
     1. [Schemas Page](https://docs.talawa.io/docs/developers/talawa-api/schemas) 
     1. [Mutations Page](https://docs.talawa.io/docs/developers/talawa-api/mutations) 
     1. [The markdown versions of those files can be found here](https://github.com/PalisadoesFoundation/talawa-docs/tree/develop/docs/developers/talawa-api). 

**Describe alternatives you've considered**
N/A

**Approach to be followed (optional)**
See above

**Additional context**
[![IMAGE ALT TEXT HERE](https://img.youtube.com/vi/ix3eSReJj2k/0.jpg)](https://www.youtube.com/watch?v=ix3eSReJj2k?t=79)

**Comment by EshaanAgg** [comment]
@palisadoes I would like to work on this issue.

**Comment by EshaanAgg** [comment]
@palisadoes Why are we creating a query to get the `Nth` pinned post? That would require us to to send one query for each of the post when being rendered on the frontend. Don't you think if would make sense to rather have queries like `pinnedPosts` which return all the posts that are pinned?

Then the check that if a post is the land one in the list can easily be handled on the frontend itself. What do you think?

**Comment by palisadoes** [comment]
The current design allows you to swipe to see the next post, and with lots of users having this information at the top of the need feed I thought it would be a worthwhile way to cut down on network traffic.

But I guess the app could get the full list on the first swipe because the pinned posts will not be visible before then. So your proposed approach could work for now

**Comment by EshaanAgg** [comment]
Getting a new post on each swipe would be too many network calls (as a user might be just browsing past them, not necessarily reading or interacting with them), and wouldn't allow for client side caching. Getting them as an array, and displaying them is a more feasible and scalable solution in my opinion.

**Comment by palisadoes** [comment]
That is what I was saying as the second option. I think the complete array approach is the better solution too for the reasons mentioned before.

**Comment by xoldyckk** [comment]
@palisadoes @DMills27 

Lemme ask some things:-

1. The user's home page newsfeed. Is that only for the current organization they're switched in for? Also does it include both events and posts or just the posts?
2. Is there a some constraint, something like only the posts that have been created in last 24 hrs will show up on user's newsfeed. Otherwise there'd be no difference between user's home page newsfeed and the organization's page newsfeed. Or is there no organization's newsfeed page alltogether?
3. I saw that pinned posts are not directly shown on the user's newsfeed. They're only shown when user clicks on pinned posts button. Can you confirm if this is really how pinned posts thingy is going to be implemented as UI in mobile and admin?
4. Who's allowed to pin/unpin posts amongst users/admins/superadmins?
5. Any maximum limit to number of pinned posts?

Answer me all this then it'd be easier to think of a solution.

**Comment by EshaanAgg** [comment]
Got it. 
A couple of implementation details that I'm taking away:
1. Instead of implementing a `pinned` field on each of the `Post` model, you want to have each `Organization` to have a new property called `pinnedPosts` which would be an array of `_id` of some post objects, which can be populated using a field level resolver.
2. I don't think so there is a need for pagination as of now, as the number of pinned posts is minimal. If the need for pagination arises, then it makes sense to keep the current implementation, and not the junction implementation as I see it.
3. I will update the `createPost` mutation to take a `pinned` field, and only apply the it's value if the user is an admin of that organization or a suderadmin.
4. I think that instead of a general mutation of `togglePost`, if makes more sense to have specific `pinPost` and `unpinPost` as it provides more verbosity to the codebase. Moreover in most of the cases the frontend would always be specific about what it wants to do to a post's pinned status (as I envision it). What do you think?
5. I would be removing the `pinnedPosts` query as it no serves no practical purpose now. I would only expose the `pinnedPostsByOrganization` query now (after checking the status of the user ofc, practically doing the some job as the resolver of the `pinnedPosts` field on the organization)

**Comment by EshaanAgg** [comment]
PS.
1. I haven't really worked with the `@auth` directive and user role verification. Would it be possible for you to give me a brief introduction to the same? Or can you point me to some files that I could use as a reference?
2. Since the proposed changes are drastically different from my initial implementation, it might take me a day or two to make these changes, and implement the tests for them as well.

Should I get started on the same? @xoldyckk

**Comment by xoldyckk** [comment]
> Got it. A couple of implementation details that I'm taking away:
> 
> 1. Instead of implementing a `pinned` field on each of the `Post` model, you want to have each `Organization` to have a new property called `pinnedPosts` which would be an array of `_id` of some post objects, which can be populated using a field level resolver.
> 2. I don't think so there is a need for pagination as of now, as the number of pinned posts is minimal. If the need for pagination arises, then it makes sense to keep the current implementation, and not the junction implementation as I see it.
> 3. I will update the `createPost` mutation to take a `pinned` field, and only apply the it's value if the user is an admin of that organization or a suderadmin.
> 4. I think that instead of a general mutation of `togglePost`, if makes more sense to have specific `pinPost` and `unpinPost` as it provides more verbosity to the codebase. Moreover in most of the cases the frontend would always be specific about what it wants to do to a post's pinned status (as I envision it). What do you think?
> 5. I would be removing the `pinnedPosts` query as it no serves no practical purpose now. I would only expose the `pinnedPostsByOrganization` query now (after checking the status of the user ofc, practically doing the some job as the resolver of the `pinnedPosts` field on the organization)

1. I was thinking of creating a seperate document which is used just for relation between an organization and a post. We'd query the database for this relation document and get all post ids which are related to a particular organization, then query and return them. What you're suggesting would need us to first look into organization's document, take pinned post ids, query and return those posts. Both are viable choices but my approach gives us more control over the relationship between them. But more control isn't needed right now. Both are good choices if the number of pinned posts is small compared to total number of posts for an organization.

2. Only peter can confirm whether the number of pinned posts will be small or not. If they're in the numbers of 10-20s then yes pagination isn't needed.

4. It doesn't make sense to expose two mutation fields that are just toggling a boolean to true and false. Besides this toggle behaviour will only be triggered when the clients can actually see on their screen whether the post is pinned or not. So, it will never happen that the client is oblivious to whether the post is pinned or not and toggling it might not be the desired behaviour.

5. I think I had already explained to you that exposing just about everything as a seperate query field is anti-pattern in graphql. There is no point of using graphql if you can't fetch related items together. Exposing every relation as a seperate query field defeats the whole purpose of graphql. You can do the same thing using a rest api. Seperate query fields should only be used when you have some complex filtering/sorting/constraint logic.

**Comment by EshaanAgg** [comment]
1. I totally see the merit in your approach. I can get started with any of them (a bit inclined towards mine though as it is more readable in my opinion, as the posts reside there with the organization, and someone just reading the `Mongo` Schema needn't be aware of a new table which stores this relationship)
2. Got it! The concern for pagination isn't much I agree, but the size does influence our designs for the data.
3. Okay! 
4. Hmm makes sense. Haven't Then a field-level resolver for these `pinnedPosts` on an organization is all we should be implementing, and it would be the burden of the client to ask for the same whenever it requires the same. 
(Still new to the GraphQL design philosophies! Thanks for your patience!)
5. Found some code that uses role-checking on users, and I think so I'll be able to implement it!

**Comment by xoldyckk** [comment]
> 1. I totally see the merit in your approach. I can get started with any of them (a bit inclined towards mine though as it is more readable in my opinion, as the posts reside there with the organization, and someone just reading the `Mongo` Schema needn't be aware of a new table which stores this relationship)
> 2. Got it! The concern for pagination isn't much I agree, but the size does influence our designs for the backend.
> 3. Okay!
> 4. Hmm makes sense. Haven't Then a field-level resolver for these `pinnedPosts` on an organization is all we should be implementing, and it would be the burden of the client to ask for the same whenever it requires the same.
>    (Still new to the GraphQL design philosophies! Thanks for your patience!)
> 5. Found some code that uses role-checking on users, so I think that I'll be able to implement the required checks!

So, basically I think if number of pinned posts are gonna be small it would be stupid to have a useless `pinned` field on every post for no reason. It'll just take database space. But if it's gonna be large then it'll be actively used so it won't be a stale field.

Yes I agree that using graphql directives for maintaining rbac is useless here. If all superadmins/admins/users access the same resolver and some incoming input would only be allowed to mutate the db object if the client is superadmin/admin the graphql directive can't be used on this mutation field. But the alternative is that the admins have to first create a post then send a seperate request to pin them. That just seems useless complication to me. Another approach is to use a seperate mutation field for the same thing but it will only allow admins/superadmins to access it. This seems even stupider for such a small task.

**Comment by EshaanAgg** [comment]
Got all your concerns and recommendations to implement the same. Thanks for the super-detailed review. I will try to get cracking on the implementation for the same and will update about my progress here.

**Comment by xoldyckk** [comment]
> Got all your concerns and recommendations to implement the same. Thanks for the super-detailed review. I will try to get cracking on the implementation for the same and will update about my progress here.

Having seperate mutation requests once for creating the post, then for pinning it is not entirely an absolute horrendous thing, just that it will make the admins do two requests instead of one. You're free to choose this way if you want cuz it'll align with the schema directives thing better.

**Comment by palisadoes** [comment]
The creation of a post and pinning it may have to be done in the same Talawa admin screen.

If posted at a peak time it may be difficult for the admin to find the post again to pin it, unless there is a "My posts" area like the profile pushes of Twitter and Instagram. 

Profile pages with "my posts" is the topic of another discussion.

**Comment by xoldyckk** [comment]
Then post creation and pinning will have to be implemented in same mutation, and it'll have to check that a non admin doesn't try to pin it. If they do a simple fix is to just ignore their input for `pinned` boolean.

**Comment by xoldyckk** [comment]
@palisadoes 

I personally think that pinning posts should be restricted to a short number of posts. Because without this restriction a pinned post loses its meaning compared to a normal post.

Regardless of this, if you're still unclear about whether it being constrained to only a short number of posts can change in the future, we can just implement a pagination technique right now. It will work for both cases, be future proof and it's not complex at all.

**Comment by palisadoes** [comment]
It'll be no more than 5-6 posts. The current design for the mobile app only allows for that.

**Comment by xoldyckk** [comment]
@palisadoes Let me rephrase the question. Do you want to put a limit on the number of pinned posts for a particular organization? Or it's gonna be wild wild west?

**Comment by palisadoes** [comment]
5-6 pinned posts pet organization at any time for now.

Hopefully we can get a look and feel like this in the near future.

This is the Economist mobile app that:

1. shows 8 total. The first four, then a swipe for the next four.
1. When you click on the thumbnail of any post, you see the whole thing.
1. You can swipe to see each of the pinned posts on full
1. Clicking the back button store gets you back to the home page

### First Four 
![Screenshot_20230302-060206_Economist.jpg](https://user-images.githubusercontent.com/16875803/222275241-587d970c-0933-452c-b66c-1042b27861d6.jpg)

### Second Four 
![Screenshot_20230302-060511_Economist.jpg](https://user-images.githubusercontent.com/16875803/222275604-cd83be26-f07d-4512-8ee9-0ce93ca96bd4.jpg)

### Click Through to Post
![Screenshot_20230302-060809_Economist.jpg](https://user-images.githubusercontent.com/16875803/222276868-ff73ca58-a2ad-40c4-aef6-c716efab3089.jpg)


### Swipe in Post

![Uploading Screenshot_20230302-060818_Economist.jpg …]()
